## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [4]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  4/28/2023   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  5/19/2023   
2                swatson8146@payne.net      57113   1902144  1/29/2023   
3               bakernathan@benson.com      46554   9031802  4/25/2023   
4  christinaschwartz9252@mcconnell.com      92089   1322274  5/28/2023   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [5]:
# View the column names in the data
column_names = df.columns
print(column_names)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [7]:
# Use the describe function to gather some basic statistics
statistics = df.describe()
print(statistics)


          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [9]:
# Use this space to do any additional research
# and familiarize yourself with the data.
data_types = df.dtypes
print(data_types)


first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object


In [10]:
# Use this space to do any additional research
# and familiarize yourself with the data.
missing_values = df.isnull().sum()
print(missing_values)

first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64


In [18]:
# What three item categories had the most entries?
# Group the data by item category and count the number of entries in each category
category_counts = df['category'].value_counts()

# Get the top three item categories with the most entries
top_three_categories = category_counts.head(3)

# Display the top three item categories with the most entries
top_three_categories


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [25]:
# For the category with the most entries,
# which subcategory had the most entries?
# Filter the DataFrame for entries where the item category is 'consumables'
consumables_df = df[df['category'] == 'consumables']

# Group the filtered data by subcategory and count the number of entries in each subcategory
subcategory_counts = consumables_df['subcategory'].value_counts()

# Get the subcategory with the most entries within the 'consumables' category
most_frequent_subcategory = subcategory_counts.idxmax()
most_frequent_subcategory


'bathroom supplies'

In [27]:
# Which five clients had the most entries in the data?
client_counts = df['client_id'].value_counts()

# Get the top five clients with the most entries
top_five_clients = client_counts.head(5)

top_five_clients


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [28]:
# Store the client ids of those top 5 clients in a list.
top_five_client_ids = top_five_clients.index.tolist()

top_five_client_ids


[33615, 66037, 46820, 38378, 24741]

In [29]:
# How many total units (the qty column) did the
# client with the most entries order order?
# Find the client with the most entries
client_with_most_entries = top_five_clients.idxmax()

# Filter the DataFrame for entries corresponding to the client with the most entries
client_df = df[df['client_id'] == client_with_most_entries]

# Sum the quantity (qty) column for the client
total_units_ordered = client_df['qty'].sum()

total_units_ordered


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [38]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']

# Display only the 'unit_price', 'qty', and 'line_subtotal' columns
output_df = df[['unit_price', 'qty', 'line_subtotal']]

output_df.head()


unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00

In [37]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
# Define a function to calculate shipping price based on weight
# Define a function to calculate total shipping price based on weight and quantity
# Create a column for total weight
df['total_weight'] = df['unit_weight'] * df['qty']

# Define a function to calculate total shipping price based on total weight
def calculate_total_shipping_price(total_weight):
    shipping_price_per_pound = 7 if total_weight > 50 else 10  # Determine shipping price per pound
    total_shipping_price = shipping_price_per_pound * total_weight  # Calculate total shipping price
    return total_shipping_price

# Create a new column 'shipping_price' by applying the function to the 'total_weight' column
df['shipping_price'] = df['total_weight'].apply(calculate_total_shipping_price)


df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']]


unit_price  unit_weight   qty  total_weight  shipping_price
0         1096.80         7.50   105        787.50         5512.50
1           24.95         1.49    21         31.29          312.90
2           13.52         1.68    39         65.52          458.64
3           36.42         1.23    29         35.67          356.70
4          195.10        46.43    20        928.60         6500.20
...           ...          ...   ...           ...             ...
54634       83.13         2.25    33         74.25          519.75
54635      206.59        11.70    47        549.90         3849.30
54636       65.66         4.16   475       1976.00        13832.00
54637        1.48        18.04   112       2020.48        14143.36
54638        3.01         2.07  1031       2134.17        14939.19

[54639 rows x 5 columns]

In [67]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
# Define sales tax rate
sales_tax_rate = 0.0925  # 9.25%

# Step 1: Calculate the subtotal for each line using the unit price and quantity
df['line_subtotal'] = df['unit_price'] * df['qty']

# Step 2: Calculate the shipping price for each line based on the total weight
df['shipping_price'] = df['total_weight'].apply(lambda weight: 7 * weight if weight > 50 else 10 * weight)

# Step 4: Calculate the total price by adding the line price and sales tax
df['line_price'] = df['line_price'] * (1 + sales_tax_rate)
df['line_price'] = df['line_price'].round(2)


df[['line_subtotal', 'shipping_price', 'line_price']]


line_subtotal  shipping_price  line_price
0          115164.00         5512.50   157357.35
1             523.95          312.90     1091.22
2             527.28          458.64     1285.60
3            1056.18          356.70     1842.34
4            3902.00         6500.20    13564.05
...              ...             ...         ...
54634        2743.29          519.75     4254.87
54635        9709.73         3849.30    17680.43
54636       31188.50        13832.00    58704.94
54637         165.76        14143.36    18658.52
54638        3103.31        14939.19    23526.70

[54639 rows x 3 columns]

In [64]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
# Step 1: Calculate the subtotal for each line using the unit cost and quantity
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  4/28/2023   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  5/19/2023   
2                swatson8146@payne.net      57113   1902144  1/29/2023   
3               bakernathan@benson.com      46554   9031802  4/25/2023   
4  christinaschwartz9252@mcconnell.com      92089   1322274  5/28/2023   

   order_week  order_year  ... line_number   subtotal shipping_price  \
0          17        2023  ...           1  115164.00        5512.50   
1          20        2023  ...           0     523.95         312.90   
2           4        2023  ...           6     527.28         458.64   
3          17        2023  ...           3    1056.18         356.70   
4          21        2023  ...           1    3902.00        6500.20   

   total_shipping_price  total_weight  line_subtotal  line_price  total_price  \
0                7875.0        787.50       80084.55   131839.08    131839.08   
1                 312.9         31.29         316.89      914.26       914.26   
2                 655.2         65.52         306.54     1077.12      1077.12   
3                 356.7         35.67         720.65     1543.57      1543.57   
4                9286.0        928.60        2163.40    11364.40     11364.40   

   line_cost  line_profit  
0   85597.05     11162.58  
1     629.79        77.41  
2     765.18        91.20  
3    1077.35       130.69  
4    8663.60       962.20  

[5 rows x 27 columns]

In [51]:
# Create a column for the profit of
# each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  4/28/2023   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  5/19/2023   
2                swatson8146@payne.net      57113   1902144  1/29/2023   
3               bakernathan@benson.com      46554   9031802  4/25/2023   
4  christinaschwartz9252@mcconnell.com      92089   1322274  5/28/2023   

   order_week  order_year  ... line_number   subtotal shipping_price  \
0          17        2023  ...           1  115164.00        5512.50   
1          20        2023  ...           0     523.95         312.90   
2           4        2023  ...           6     527.28         458.64   
3          17        2023  ...           3    1056.18         356.70   
4          21        2023  ...           1    3902.00        6500.20   

   total_shipping_price  total_weight  line_subtotal  line_price  total_price  \
0                7875.0        787.50      115164.00   131839.08    131839.08   
1                 312.9         31.29         523.95      914.26       914.26   
2                 655.2         65.52         527.28     1077.12      1077.12   
3                 356.7         35.67        1056.18     1543.57      1543.57   
4                9286.0        928.60        3902.00    11364.40     11364.40   

   line_cost  line_profit  
0  120676.50     11162.58  
1     836.85        77.41  
2     985.92        91.20  
3    1412.88       130.69  
4   10402.20       962.20  

[5 rows x 27 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [65]:
# Check your work using the totals above
# Calculate the total price for each order ID





## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.



33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [17]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.



client_id     qty  shipping_price   line_price    line_cost  line_profit
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.



In [19]:
# Sort the updated data by "Total Profit" form highest to lowest


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M